Added setuptools structure to enable installation

rikwade opened the pull request:
Added setup.py for setuptools installation support using Pip
Changed pico_generator to include a main() function for script execution when installed
GIF and TSV copied to installation directory
Changed pico_generator to read TSV file from current installation directory
Added section to README for Pip installation

rikwade committed:
Added setup.py to enable installation to user or system directory using setuptools
Re-structured directory tree to permit setuptools installation
Added "main" function definition in pico_project.py to permit executable installation using setuptools
Updated pico_configs.tsv configuration file read function to read from python script installation directory

JamesH65 commented:
@lurch These changes look sane to me but I am not a Python expert, any comments from you?

JamesH65 commented on README.md:
```diff
@@ -42,14 +42,21 @@ optional arguments:
 
 You can add specific features to your project by selecting them from the check boxes on the GUI. This will ensure the build system adds the appropriate code to the build, and also adds simple example code to the project showing how to use the feature. There are a number of options available, which provide the following functionality.
 
+## Installation using Python Pip (Linux)
```
This section is in the wrong place in the readme. Please rebase on the latest readme, then I think probably best to move this to the bottom of the page?

lurch replied:
Probably worth clarifying that this is optional too, as it isn't how we (currently) document usage of pico-project-generator.

lurch commented:
Sorry, I hadn't seen this before, oops. The changes to `pico_project.py` look good (although you'll obviously need to rebase onto the latest changes), but I'm afraid I've never written a `setup.py` script. Don't most projects that use `pip` use it to install stuff from PyPI? Is it absolutely necessary to move the repo-files into a subdirectory?

rikwade commented:
> Don't most projects that use `pip` use it to install stuff from PyPI? Is it absolutely necessary to move the repo-files into a subdirectory?

Yes, making the package available on PyPi is also possible. Ref: https://packaging.python.org/tutorials/packaging-projects/. Structure is similar to the one I'm making.

I would have preferred to have kept the files in the root directory. It seems that having a subdirectory with the Python package (application in our case) name is "standard". Refs:
https://docs.python-guide.org/writing/structure/
https://realpython.com/python-application-layouts/

lurch commented:
Fair enough, always good to adhere to standards :slightly_smiling_face:

rikwade committed:
Moved Pip installation information to the bottom of README and marked as
optional.

rikwade commented:
> Sorry, I hadn't seen this before, oops. The changes to `pico_project.py` look good (although you'll obviously need to rebase onto the latest changes), but I'm afraid I've never written a `setup.py` script. Don't most projects that use `pip` use it to install stuff from PyPI? Is it absolutely necessary to move the repo-files into a subdirectory?

Have rebased as requested and put pip installation as "optional" and at the foot of the README.

lurch commented on README.md:
```diff
@@ -77,10 +75,15 @@ Create VSCode Project | As well as the CMake files, also create the appropriate
 
+This application can **optionally** be installed to your system using standard Python installation tools. If you would like to do so, follow the instructions below. If you do not wish to install the application, you can run it from the the pico_project directory.
 
+Clone the repository with: `$ git clone https://github.com/raspberrypi/pico-project-generator` (assuming you are using HTTP to clone repositories). Change to the repository's directory with: `$ cd pico-project-generator`.
```
I don't think this "(assuming you are using HTTP to clone repositories)" bit is necessary?

lurch commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # pico-project-generator
 
-This is a command line or GUI tool, written in Python, to automatically generate a Pico C SDK Project.
+This is a command line or GUI tool, written in Python, to automatically generate a Raspberry Pi Pico C SDK Project.
```
Perhaps this should now say C / C++ after #23 :wink:

lurch commented on pico_project/pico_project.py:
```diff
@@ -1101,6 +1101,9 @@ def generateProjectFiles(projectPath, projectName, sdkPath, projects, debugger):
 def LoadConfigurations():
+    # open the pico_configs.tsv from the executable's directory
+    this_module_location = os.path.dirname(os.path.realpath(__file__))
+    tsv_file = os.path.join(this_module_location, "pico_configs.tsv")
```
This bit probably needs updating in light of the new `--tsv` option. Note how the line below does `open(args.tsv)` and not `open(tsv_file)`. I suspect the correct fix might be to move the `this_module_location ...` stuff into `ParseCommandLine` ?

rikwade replied:
Done as suggested. Tested works OK.

rikwade committed:
Updated README with minor changes based on PR feedback.
Updated pico_project to move default tsv file specification to argparse
function.

lurch commented on pico_project/pico_project.py:
```diff
@@ -1101,6 +1105,7 @@ def generateProjectFiles(projectPath, projectName, sdkPath, projects, debugger):
 
 
 def LoadConfigurations():
+    print("tsv: ",args.tsv)
```
Leftover debugging?

rikwade committed:
remove debug
